k8s/k8s.go

```diff
@@ -349,6 +397,53 @@ func (k *k8sOps) RemoveLabelOnNode(name, key string) error {
 	return err
 }
 
+// NodeWatchFunc is a callback provided to the WatchNode function
+// which is invoked when the v1.Node object is changed.
+type NodeWatchFunc func(node *v1.Node) error
+
+func (k *k8sOps) WatchNode(node *v1.Node, watchNodeFn NodeWatchFunc) error {
+	if node == nil {
+		return fmt.Errorf("no node given to watch")
+	}
+
+	if err := k.initK8sClient(); err != nil {
+		return err
+	}
+
+	nodeHostname, has := node.GetLabels()[hostnameKey]
+	if !has || nodeHostname == "" {
+		return fmt.Errorf("no hostname label")
+	}
+
+	selector := hostnameKey + "==" + nodeHostname
+	listOptions := meta_v1.ListOptions{
+		Watch:         true,
+		LabelSelector: selector,
+	}
+	watchInterface, err := k.client.Core().Nodes().Watch(listOptions)
+	if err != nil {
+		return err
+	}
+
+	// fire off watch function
+	go func() {
+		for {
+			select {
+			case event, more := <-watchInterface.ResultChan():
+				if !more {
+					// log.Warn("Kubernetes node watch channel closed")
```

Comment:
Closing of the watch channel needs to be handled since this has been seen on QA setups very frequently. Retries were added in porx to perform this watch. Referring to below function in porx k8s_node_utils.go

You can check this version in and add the retries in a subsequent commit with a TODO here.

```
func (knu *k8sNodeUtils) handleWatch(
	watchInterface k8s_watch.Interface,
	watchNodeFn NodeWatchFunc,
	nodeHostname string,
) {
	for {
		select {
		case event, more := <-watchInterface.ResultChan():
			if !more {
				var err error
				for retries := 0; retries < maxKubernetesWatchRetries; retries++ {
					// We lost the watch connection most probably because kube api-server is down.
					// Sleep for sometime and retry again.
					time.Sleep(5 * time.Second)
					err = knu.WatchNode(nodeHostname, watchNodeFn)
					if err != nil {
						continue
					}
					break
				}
				if err != nil {
					logrus.Warn("Kubernetes node watch "+
						"channel closed. Failed to restart watch. Error: %v", err)

				}
				return
			}
			k8sNode, ok := event.Object.(*k8s_types.Node)
			if ok {
				watchNodeFn(k8sNode)
			}
		}
	}
}
```

Reply:
Why can't we check in this version "as is" -- and have Aditya fix up the function properly as he is the original author.  I just ported the function from API v1.5 to v4.0.

Reply:
Sure, go ahead.